DICOMPlugins/DICOMM3DPlugin.py

```diff
@@ -0,0 +1,641 @@
+from __future__ import absolute_import
+import glob
+import os
+import json
+import vtk
+import io
+import shutil
+import vtkSegmentationCorePython as vtkSegmentationCore
+import logging
+import pydicom
+
+from base.DICOMPluginBase import DICOMPluginBase
+
+import slicer
+from DICOMLib import DICOMLoadable
+
+from SlicerDevelopmentToolboxUtils.mixins import ModuleLogicMixin
+from SlicerDevelopmentToolboxUtils.constants import DICOMTAGS
+from six.moves import range
+
+
+#
+# This is the plugin to handle translation of DICOM M3D objects
+# M3D stands for Model for 3D Manufacturing.
+# DICOM M3D objects can be for example .STL files encoded into a DICOM M3D object.
+#
+class DICOMM3DPluginClass(DICOMPluginBase):
+
+  def __init__(self):
+    super(DICOMM3DPluginClass,self).__init__()
+    self.loadType = "DICOMM3D"
+
+  def examineFiles(self,files):
+    """ Returns a list of DICOMLoadable instances
+    corresponding to ways of interpreting the
+    files parameter.
+    """
+    loadables = []
+
+    for candidateFile in files:
+
+      uid = slicer.dicomDatabase.fileValue(candidateFile, self.tags['instanceUID'])
+      if uid == '':
+        return []
+
+      desc = slicer.dicomDatabase.fileValue(candidateFile, self.tags['seriesDescription'])
+      if desc == '':
+        desc = "Unknown"
+
+      #read EncapsulatedDocument for candidateFile
+      docFile = self.getEncapsulatedDocument(candidateFile)
+
+      #read EncapsulatedDocumentLength for candidateFile
+      docLengthFile = self.getEncapsulatedDocumentLength(candidateFile)
+
+      #read modality type to flag M3D object.
+      isDicomM3D = (slicer.dicomDatabase.fileValue(candidateFile, self.tags['modality']) == 'M3D')
```

Comment:
This should go first, so you can quickly bail out if the modality is not M3D.  Otherwise you will be reading every candidate file (which could be a giant series of thousands of files) over and over again!